pkg/etcd/etcd.go

```diff
@@ -1087,11 +1088,16 @@ func (e *ETCD) manageLearners(ctx context.Context) {
 
 		for _, member := range members.Members {
 			var node *v1.Node
-			for _, node = range nodes {
-				if strings.HasPrefix(member.Name, node.Name+"-") {
-					nodeIsMember[node.Name] = true
+			for _, n := range nodes {
+				if strings.HasPrefix(member.Name, n.Name+"-") {
+					node = n
+					break
 				}
 			}
+			if node == nil {
+				continue
+			}
+			nodeIsMember[node.Name] = true
 
 			if member.IsLearner {
 				if err := e.trackLearnerProgress(ctx, progress, member); err != nil {
```

Comment:
1. Move these bits into a separate block up top, before `var node *v1.Node`. This will keep learner progress tracking and promotion separate from status condition updates.

2. Move the Learner/Unhealthy/Voter status updates into an if/else-if/else block down below. The code for all three is pretty much identical, and could be much reduced if you determined the desired status first, and then only had a single call to `e.setEtcdStatusCondition`